// @mikro-orm/sql-highlighter@1.0.1/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOKEN_TYPE_TO_HIGHLIGHT = exports.HighlightSubject = exports.TokenType = void 0;
var TokenType;
(function (TokenType) {
    TokenType[TokenType["WHITESPACE"] = 0] = "WHITESPACE";
    TokenType[TokenType["WORD"] = 1] = "WORD";
    TokenType[TokenType["QUOTE"] = 2] = "QUOTE";
    TokenType[TokenType["BACKTICK_QUOTE"] = 3] = "BACKTICK_QUOTE";
    TokenType[TokenType["RESERVED"] = 4] = "RESERVED";
    TokenType[TokenType["RESERVED_TOPLEVEL"] = 5] = "RESERVED_TOPLEVEL";
    TokenType[TokenType["RESERVED_NEWLINE"] = 6] = "RESERVED_NEWLINE";
    TokenType[TokenType["BOUNDARY"] = 7] = "BOUNDARY";
    TokenType[TokenType["COMMENT"] = 8] = "COMMENT";
    TokenType[TokenType["BLOCK_COMMENT"] = 9] = "BLOCK_COMMENT";
    TokenType[TokenType["NUMBER"] = 10] = "NUMBER";
    TokenType[TokenType["ERROR"] = 11] = "ERROR";
    TokenType[TokenType["VARIABLE"] = 12] = "VARIABLE";
    TokenType[TokenType["LITERAL"] = 13] = "LITERAL";
    TokenType[TokenType["BUILT_IN"] = 14] = "BUILT_IN";
})(TokenType = exports.TokenType || (exports.TokenType = {}));
var HighlightSubject;
(function (HighlightSubject) {
    HighlightSubject["BOUNDARY"] = "boundary";
    HighlightSubject["WORD"] = "word";
    HighlightSubject["BACKTICK_QUOTE"] = "backtickQuote";
    HighlightSubject["QUOTE"] = "quote";
    HighlightSubject["RESERVED"] = "reserved";
    HighlightSubject["NUMBER"] = "number";
    HighlightSubject["VARIABLE"] = "variable";
    HighlightSubject["COMMENT"] = "comment";
    HighlightSubject["FUNCTIONS"] = "functions";
    HighlightSubject["LITERAL"] = "literal";
    HighlightSubject["BUILT_IN"] = "builtIn";
})(HighlightSubject = exports.HighlightSubject || (exports.HighlightSubject = {}));
exports.TOKEN_TYPE_TO_HIGHLIGHT = {
    [TokenType.BOUNDARY]: HighlightSubject.BOUNDARY,
    [TokenType.WORD]: HighlightSubject.WORD,
    [TokenType.BACKTICK_QUOTE]: HighlightSubject.BACKTICK_QUOTE,
    [TokenType.QUOTE]: HighlightSubject.QUOTE,
    [TokenType.RESERVED]: HighlightSubject.RESERVED,
    [TokenType.RESERVED_TOPLEVEL]: HighlightSubject.RESERVED,
    [TokenType.RESERVED_NEWLINE]: HighlightSubject.RESERVED,
    [TokenType.NUMBER]: HighlightSubject.NUMBER,
    [TokenType.VARIABLE]: HighlightSubject.VARIABLE,
    [TokenType.COMMENT]: HighlightSubject.COMMENT,
    [TokenType.BLOCK_COMMENT]: HighlightSubject.COMMENT,
    [TokenType.LITERAL]: HighlightSubject.LITERAL,
    [TokenType.BUILT_IN]: HighlightSubject.BUILT_IN,
};
