// @mikro-orm/sql-highlighter/SqlHighlighter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqlHighlighter = void 0;
const ansi_colors_1 = __importDefault(require("ansi-colors"));
const Tokenizer_1 = require("./Tokenizer");
const enums_1 = require("./enums");
class SqlHighlighter {
    constructor(theme = {}) {
        this.theme = theme;
        this.tokenizer = new Tokenizer_1.Tokenizer();
        this.theme = Object.assign(Object.assign({}, SqlHighlighter.DEFAULT_THEME), this.theme);
    }
    /**
     * Add syntax highlighting to a SQL string
     */
    highlight(sql) {
        const tokens = this.tokenizer.tokenize(sql);
        let token;
        let ret = '';
        let position = 0;
        // eslint-disable-next-line no-cond-assign
        while (token = tokens[position++]) {
            ret += this.highlightToken(token.type, token.value);
        }
        return ret;
    }
    highlightToken(type, value) {
        if (type === enums_1.TokenType.BOUNDARY && ['(', ')'].includes(value)) {
            return value;
        }
        return this.colorize(type, value);
    }
    colorize(type, value) {
        if (!enums_1.TOKEN_TYPE_TO_HIGHLIGHT[type] || !this.theme[enums_1.TOKEN_TYPE_TO_HIGHLIGHT[type]]) {
            return value;
        }
        return this.theme[enums_1.TOKEN_TYPE_TO_HIGHLIGHT[type]](value);
    }
}
exports.SqlHighlighter = SqlHighlighter;
SqlHighlighter.DEFAULT_THEME = {
    [enums_1.HighlightSubject.QUOTE]: ansi_colors_1.default.yellow,
    [enums_1.HighlightSubject.BACKTICK_QUOTE]: ansi_colors_1.default.yellow,
    [enums_1.HighlightSubject.RESERVED]: ansi_colors_1.default.white.bold,
    [enums_1.HighlightSubject.BOUNDARY]: ansi_colors_1.default.reset,
    [enums_1.HighlightSubject.NUMBER]: ansi_colors_1.default.green,
    [enums_1.HighlightSubject.WORD]: undefined,
    [enums_1.HighlightSubject.COMMENT]: ansi_colors_1.default.green.dim,
    [enums_1.HighlightSubject.VARIABLE]: ansi_colors_1.default.yellow,
    [enums_1.HighlightSubject.FUNCTIONS]: ansi_colors_1.default.green.bold,
    [enums_1.HighlightSubject.BUILT_IN]: ansi_colors_1.default.cyan,
    [enums_1.HighlightSubject.LITERAL]: ansi_colors_1.default.cyan,
};
